src/service/resource.js

```diff
@@ -560,6 +566,32 @@ export class Resource {
   }
 
   /**
+   * @return {!Promise} resolves when underlying element is built and within the
+   *    range specified by implementation's renderOutsideViewport
+   */
+  whenWithinRenderOutsideViewport() {
+    if (this.renderOutsideViewportPromise_) {
+      return this.renderOutsideViewportPromise_;
+    }
+    return this.renderOutsideViewportPromise_ = new Promise(resolver => {
+      this.renderOutsideViewportResolve_ = resolver;
+    });
+  }
+
+  /**
+   * @private resolves render outside viewport promise if one was created via
+   *    whenWithinRenderOutsideViewport.
+   */
+  resolveRenderOutsideViewport_() {
```

Comment:
Ok, I'm clearly still tired, because the logic of this is eluding me.  I think the idea is that `whenWithinRenderOutsideViewport()` returns a promise that will block until somebody calls `renderOutsideViewportResolve_` -- is that right? Could you add some comments describing how this mechanism works?  I'm also concerned that there might be a way to get to deadlock.  I don't *think* so, by looking at this, but I can't be sure.  If there's any way that `renderOutsideViewportPromise_` can get cleared before being resolved, then you could get into a state where somebody is blocked on that promise, but nobody has a handle to it to resolve it.  Not sure that that can really happen, though.

Gah.  Reasoning about multiple execution is bad enough, without adding "no locking primitives" into the mix.